### ISSUE TITLE
ValueError: math domain error

### ISSUE THREAD

**Issue by Justheretoreportanissue** [edekeijzer/OctoPrint-WebDavBackup#16]
OctoPrint v1.5.3, Python v3.7.3, OctoPi v0.18.0

Fresh install on Raspberry 4

WebDAV server OwnCloud v10.5.0.10 with total data size of 7,8T, used 4,8T and available 2,6T

Could it be that my OwnCloud reports storage space that is too large to compute? If so, could this be handled?

`021-02-11 18:32:51,903 - octoprint.plugin - ERROR - Error while calling plugin webdavbackup`
`Traceback (most recent call last):`
`   File "/home/pi/oprint/lib/python3.7/site-packages/octoprint/plugin/__init__.py", line 271, in call_plugin`
`    result = getattr(plugin, method)(*args, **kwargs)`
`   File "/home/pi/oprint/lib/python3.7/site-packages/octoprint/util/__init__.py", line 1890, in wrapper`
`    return f(*args, **kwargs)`
`   File "/home/pi/oprint/lib/python3.7/site-packages/octoprint_webdavbackup/__init__.py", line 109, in on_event`
`    self._logger.info("Free space on server: " + _convert_size(dav_free))`
`   File "/home/pi/oprint/lib/python3.7/site-packages/octoprint_webdavbackup/__init__.py", line 46, in _convert_size`
`    i = int(math.floor(math.log(size_bytes, 1024)))`
`ValueError: math domain error`

Regards,

   /Jonas...

**Comment by AndyPC54** [comment]
I get a similar error talking to my local Synology server.

Raspberry Pi model 4
OctoPriint 1.5.3
Python 3.7.3
WebDAV Backup (0.2.1) = /home/pi/oprint/lib/python3.7/site-packages/octoprint_webdavbackup

Synology DSM 6.2.3-25426 Update 3
Backup volume available capacity 726.56 GB out of 9.38 TB

2021-02-12 12:35:44,264 - octoprint.plugins.backup - INFO - Creating backup zip at /home/pi/.octoprint/data/backup/.octoprint-backup-20210212-123544.zip (excluded: timelapse,uploads,timelapse_tmp)...
2021-02-12 12:35:44,897 - tornado.access - WARNING - 409 GET /api/printer (127.0.0.1) 5.79ms
2021-02-12 12:35:45,092 - octoprint.plugins.backup - INFO - ... done creating backup zip.
2021-02-12 12:35:45,097 - octoprint.plugins.webdavbackup - INFO - Backup /home/pi/.octoprint/data/backup/octoprint-backup-20210212-123544.zip created, will now attempt to upload to http://xxx.xxx.local/
2021-02-12 12:35:45,337 - py.warnings - WARNING - /home/pi/oprint/lib/python3.7/site-packages/urllib3/connectionpool.py:988: InsecureRequestWarning: Unverified HTTPS request is being made to host 'xxx.xxx.local'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
  InsecureRequestWarning,

2021-02-12 12:35:45,683 - octoprint.plugin - ERROR - Error while calling plugin webdavbackup
Traceback (most recent call last):
  File "/home/pi/oprint/lib/python3.7/site-packages/octoprint/plugin/__init__.py", line 271, in call_plugin
    result = getattr(plugin, method)(*args, **kwargs)
  File "/home/pi/oprint/lib/python3.7/site-packages/octoprint/util/__init__.py", line 1890, in wrapper
    return f(*args, **kwargs)
  File "/home/pi/oprint/lib/python3.7/site-packages/octoprint_webdavbackup/__init__.py", line 109, in on_event
    self._logger.info("Free space on server: " + _convert_size(dav_free))
  File "/home/pi/oprint/lib/python3.7/site-packages/octoprint_webdavbackup/__init__.py", line 46, in _convert_size
    i = int(math.floor(math.log(size_bytes, 1024)))
**TypeError: must be real number, not str**


Regards,
Andy

**Comment by Justheretoreportanissue** [comment]
Hi!

Because I have so little to do on a Friday afternoon I decided to do some quick and dirty troubleshooting.

If I comment out lines 109-111 and line 113, replacing 113 with this in `__init__.py`:
`if 0 > 1:`

All works fine and I get my backup sent to my OwnCloud in the correct directory.

I'll bet my pants on that `def _convert_size(size_bytes):` is the problem.

I'm not too skilled in, python I believe, but someone should be able to figure out how to sort this out.

On a second note I think we might have a problem with large numbers here.
If I bypass the `def _convert_size(size_bytes):` and just modify lines 109 and 111 to something like this:
`self._logger.info("Free space on server: " + str(dav_free))`

I then get this in the log (backup still working since `if 0 > 1:`):
`2021-02-12 15:51:04,371 - octoprint.plugins.webdavbackup - INFO - Free space on server: -3`
`2021-02-12 15:51:04,372 - octoprint.plugins.webdavbackup - INFO - Backup file size: 912261`

My free space is -3 (!?)

Please have a look at the code and see if something can be done to it, preferably before I destroy my whole weekend over this and force myself to learn python too.

Regards,

   /Jonas...

**Comment by AndyPC54** [comment]
I have done a few more tests with the connection to my Synology.  I have corrected the connection string and now get different errors. Using the Linux 'cadaver' commandline WebDav client on the octopi I can successfully attach to the Synology and can 'put' a backup.zip file into the home directory for the account I am using.

It looks like the Synology WebDav server does not support the 'free' method.
This is the error stack I get when creating a backup:

2021-02-12 17:47:18,309 - octoprint.plugins.backup - INFO - Creating backup zip at /home/pi/.octoprint/data/backup/.octoprint-backup-20210212-174718.zip (excluded: timelapse,uploads,timelapse_tmp)...
2021-02-12 17:47:19,054 - octoprint.plugins.backup - INFO - ... done creating backup zip.
2021-02-12 17:47:19,060 - octoprint.plugins.webdavbackup - INFO - Backup /home/pi/.octoprint/data/backup/octoprint-backup-20210212-174718.zip created, will now attempt to upload to https://xxx.xxx.local:5006/
2021-02-12 17:47:19,096 - py.warnings - WARNING - /home/pi/oprint/lib/python3.7/site-packages/urllib3/connectionpool.py:988: InsecureRequestWarning: Unverified HTTPS request is being made to host 'xxx.xxx.local'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
  InsecureRequestWarning,

2021-02-12 17:47:19,192 - octoprint.plugin - ERROR - Error while calling plugin webdavbackup
Traceback (most recent call last):
  File "/home/pi/oprint/lib/python3.7/site-packages/webdav3/client.py", line 944, in parse_free_space_response
    return int(node.text)
TypeError: int() argument must be a string, a bytes-like object or a number, not 'NoneType'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/pi/oprint/lib/python3.7/site-packages/octoprint_webdavbackup/__init__.py", line 78, in on_event
    dav_free = davclient.free()
  File "/home/pi/oprint/lib/python3.7/site-packages/webdav3/client.py", line 66, in _wrapper
    res = fn(self, *args, **kw)
  File "/home/pi/oprint/lib/python3.7/site-packages/webdav3/client.py", line 277, in free
    return WebDavXmlUtils.parse_free_space_response(response.content, self.webdav.hostname)
  File "/home/pi/oprint/lib/python3.7/site-packages/webdav3/client.py", line 948, in parse_free_space_response
    raise MethodNotSupported(name='free', server=hostname)
**webdav3.exceptions.MethodNotSupported: Method 'free' not supported for https://xxx.xxx.local:5006**

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/pi/oprint/lib/python3.7/site-packages/octoprint/plugin/__init__.py", line 271, in call_plugin
    result = getattr(plugin, method)(*args, **kwargs)
  File "/home/pi/oprint/lib/python3.7/site-packages/octoprint/util/__init__.py", line 1890, in wrapper
    return f(*args, **kwargs)
  File "/home/pi/oprint/lib/python3.7/site-packages/octoprint_webdavbackup/__init__.py", line 106, in on_event
    self._logger.error("An unexpected WebDAV error was encountered: " + exception.args)
TypeError: can only concatenate str (not "tuple") to str


I see that you are planning on providing enhancements to test connections and to show error messages in the UI.  Due to the wide variety of WebDav servers available these features will be very useful for debugging.

Thanks,
Andy

**Comment by lightmaster** [comment]
Adding in that I have the same Error why trying to connect to a Synology WebDAV server:
```
2021-02-16 03:46:03,426 - octoprint.plugins.webdavbackup - INFO - Backup /home/pi/.octoprint/data/backup/octoprint-backup-20210216-034557.zip created, will now attempt to upload to https://10.0.0.14:5006
2021-02-16 03:46:03,508 - py.warnings - WARNING - /home/pi/oprint/lib/python3.7/site-packages/urllib3/connectionpool.py:1020: InsecureRequestWarning: Unverified HTTPS request is being made to host '10.0.0.14'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
  InsecureRequestWarning,

2021-02-16 03:46:03,879 - octoprint.plugin - ERROR - Error while calling plugin webdavbackup
Traceback (most recent call last):
  File "/home/pi/oprint/lib/python3.7/site-packages/webdav3/client.py", line 944, in parse_free_space_response
    return int(node.text)
TypeError: int() argument must be a string, a bytes-like object or a number, not 'NoneType'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/pi/oprint/lib/python3.7/site-packages/octoprint_webdavbackup/__init__.py", line 78, in on_event
    dav_free = davclient.free()
  File "/home/pi/oprint/lib/python3.7/site-packages/webdav3/client.py", line 66, in _wrapper
    res = fn(self, *args, **kw)
  File "/home/pi/oprint/lib/python3.7/site-packages/webdav3/client.py", line 277, in free
    return WebDavXmlUtils.parse_free_space_response(response.content, self.webdav.hostname)
  File "/home/pi/oprint/lib/python3.7/site-packages/webdav3/client.py", line 948, in parse_free_space_response
    raise MethodNotSupported(name='free', server=hostname)
webdav3.exceptions.MethodNotSupported: Method 'free' not supported for https://10.0.0.14:5006

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/pi/oprint/lib/python3.7/site-packages/octoprint/plugin/__init__.py", line 271, in call_plugin
    result = getattr(plugin, method)(*args, **kwargs)
  File "/home/pi/oprint/lib/python3.7/site-packages/octoprint/util/__init__.py", line 1890, in wrapper
    return f(*args, **kwargs)
  File "/home/pi/oprint/lib/python3.7/site-packages/octoprint_webdavbackup/__init__.py", line 106, in on_event
    self._logger.error("An unexpected WebDAV error was encountered: " + exception.args)
TypeError: can only concatenate str (not "tuple") to str
2021-02-16 03:47:13,503 - octoprint.server.util.sockjs - INFO - Client connection closed: 4.21.25.50
```

**Comment by bplein** [comment]
Trying to use WebDav Backup plugin for the first time (so I've had no success so far). Math domain error for me as well. 



```
2021-02-19 20:36:02,322 - octoprint.plugins.webdavbackup - DEBUG - Filename for upload: octoprint-20210219.zip
2021-02-19 20:36:03,383 - octoprint.plugin - ERROR - Error while calling plugin webdavbackup
Traceback (most recent call last):
  File "/home/pi/oprint/lib/python3.7/site-packages/octoprint/plugin/__init__.py", line 271, in call_plugin
    result = getattr(plugin, method)(*args, **kwargs)
  File "/home/pi/oprint/lib/python3.7/site-packages/octoprint/util/__init__.py", line 1890, in wrapper
    return f(*args, **kwargs)
  File "/home/pi/oprint/lib/python3.7/site-packages/octoprint_webdavbackup/__init__.py", line 109, in on_event
    self._logger.info("Free space on server: " + _convert_size(dav_free))
  File "/home/pi/oprint/lib/python3.7/site-packages/octoprint_webdavbackup/__init__.py", line 46, in _convert_size
    i = int(math.floor(math.log(size_bytes, 1024)))
ValueError: math domain error
```
Here is my environment

```
browser.user_agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0.3 Safari/605.1.15
connectivity.connection_check: 9.9.9.9:53
connectivity.connection_ok: true
connectivity.enabled: false
connectivity.online: true
connectivity.resolution_check: octoprint.org
connectivity.resolution_ok: true
env.hardware.cores: 4
env.hardware.freq: 1400
env.hardware.ram: 782680064
env.os.bits: 32
env.os.id: linux
env.os.platform: linux
env.plugins.pi_support.model: Raspberry Pi 3 Model B Plus Rev 1.3
env.plugins.pi_support.octopi_version: 0.18.0
env.plugins.pi_support.throttle_state: 0x80000
env.python.pip: 20.3.3
env.python.version: 3.7.3
env.python.virtualenv: true
octoprint.safe_mode: false
octoprint.version: 1.5.3
printer.firmware: Prusa-Firmware-Buddy 4.2.1-final+2072 (Github)

```

**Comment by edekeijzer** [comment]
@Justheretoreportanissue thanks for your investigations. The -3 value is quite weird, I'll build in an option to print the raw value returned by the webdav.free command.
@AndyPC54 it appears your WebDAV server does not support retrieving free space. 
Seeing these various issues related to the space check, it seems like a lot can go wrong while checking space on the server - not being supported, not returning correct values. I think the easiest thing to do is to make checking the space optional and work on some proper error handling afterwards.

**Comment by edekeijzer** [comment]
I have just released 0.3.0 which allows you to disable the free space checking, as well as some bugfixes to make the check more reliable. Please set the update channel for my plug-in to 'Release Candidate' in the Software Update settings.

**Comment by AndyPC54** [comment]
I have successfully installed the release candidate 0.3.0.

I getting a different error when connecting with the 'Check available space' option cleared.

2021-03-04 19:51:53,386 - octoprint.plugins.backup - INFO - Creating backup zip at /home/pi/.octoprint/data/backup/.octoprint-backup-20210304-195153.zip (excluded: timelapse,uploads,timelapse_tmp)...
2021-03-04 19:51:54,394 - octoprint.plugins.backup - INFO - ... done creating backup zip.
2021-03-04 19:51:54,399 - octoprint.plugins.webdavbackup - INFO - Backup /home/pi/.octoprint/data/backup/octoprint-backup-20210304-195153.zip created, will now attempt to upload to http://xxx.xxx.xxx:5005
2021-03-04 19:51:54,509 - octoprint.plugins.webdavbackup - ERROR - Server did not return WebDAV root, something is probably wronkg with your settings.

Note there is a spelling error in the error message 'wronkg' which allowed me to track the code down to the '__init__.py' script.

I have modified the code to ignore the two failed 'WebDAV root' check statuses and can now get the following stack:

2021-03-04 21:51:38,301 - octoprint.plugins.backup - INFO - Creating backup zip at /home/pi/.octoprint/data/backup/.octoprint-backup-20210304-215138.zip (excluded: timelapse,uploads,timelapse_tmp)...
2021-03-04 21:51:39,287 - octoprint.plugins.backup - INFO - ... done creating backup zip.
2021-03-04 21:51:39,292 - octoprint.plugins.webdavbackup - INFO - Backup /home/pi/.octoprint/data/backup/octoprint-backup-20210304-215138.zip created, will now attempt to upload to http://ds918.router.local:5005/
2021-03-04 21:51:39,313 - octoprint.plugins.webdavbackup - ERROR - Server did not return WebDAV root, something is probably wronkg with your settings.
2021-03-04 21:51:39,313 - octoprint.plugins.webdavbackup - INFO - Backup file size: 7.16 MB
2021-03-04 21:51:39,444 - octoprint.plugins.webdavbackup - ERROR - Could not find WebDAV root, something is probably wrong with your settings.
2021-03-04 21:51:39,459 - octoprint.plugin - ERROR - Error while calling plugin webdavbackup
Traceback (most recent call last):
  File "/home/pi/oprint/lib/python3.7/site-packages/octoprint/plugin/__init__.py", line 271, in call_plugin
    result = getattr(plugin, method)(*args, **kwargs)
  File "/home/pi/oprint/lib/python3.7/site-packages/octoprint/util/__init__.py", line 1890, in wrapper
    return f(*args, **kwargs)
  File "/home/pi/oprint/lib/python3.7/site-packages/octoprint_webdavbackup/__init__.py", line 157, in on_event
    if _recursive_create_path(upload_path):
  File "/home/pi/oprint/lib/python3.7/site-packages/octoprint_webdavbackup/__init__.py", line 149, in _recursive_create_path
    if _recursive_create_path(ospath.abspath(ospath.join(path, ".."))):
  File "/home/pi/oprint/lib/python3.7/site-packages/octoprint_webdavbackup/__init__.py", line 150, in _recursive_create_path
    davclient.mkdir(path)
  File "/home/pi/oprint/lib/python3.7/site-packages/webdav3/client.py", line 66, in _wrapper
    res = fn(self, *args, **kw)
  File "/home/pi/oprint/lib/python3.7/site-packages/webdav3/client.py", line 313, in mkdir
    raise RemoteParentNotFound(directory_urn.path())
webdav3.exceptions.RemoteParentNotFound: Remote parent for: /homes/ not found

The home directory for the user is '/homes/octobackup'.  I believe there is a restriction with the Synology WebDAV server that no-admin users may only use WebDAV to their home directory.  However from the command line cadaver program I can 'ls /' and get a valid listing.

I have tried researching the Synology server capabilities but haven't found any detail.

Do you want me to raise this against this original thread or as a new issue against 0.3.0?  I am very new to working with GitHub.

Regards,
Andy

**Comment by lightmaster** [comment]
Same error on 0.3.0 RC

```
2021-03-04 17:55:35,846 - octoprint.plugins.webdavbackup - INFO - Backup /home/pi/.octoprint/data/backup/octoprint-backup-20210304-175527.zip created, will now attempt to upload to https://10.0.0.14:5006
2021-03-04 17:55:35,897 - py.warnings - WARNING - /home/pi/oprint/lib/python3.7/site-packages/urllib3/connectionpool.py:1020: InsecureRequestWarning: Unverified HTTPS request is being made to host '10.0.0.14'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
  InsecureRequestWarning,

2021-03-04 17:55:35,908 - octoprint.plugins.webdavbackup - ERROR - Server did not return WebDAV root, something is probably wronkg with your settings.
2021-03-04 17:56:00,118 - octoprint.server.util.sockjs - INFO - Client connection closed: 4.21.25.50
```

**Comment by Justheretoreportanissue** [comment]
Just confirming v0.3.0 solve my problem with Owncloud.
Backup now works and shows up on my server correctly.

Thanks,

   /Jonas...
